### PR TITLE
[Merged by Bors] - feat(NumberTheory/ZetaFunction): residue of Riemann zeta at s = 1

### PR DESCRIPTION
**Pull request by loefflerd**
We add the computation of the residue of the completed and non-completed Riemann zeta functions at `s = 1`. 

---

See [this zulip thread](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/Residue.20of.20the.20Riemann.20zeta.20function).

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/ZetaFunction.lean [hunk 1]
```diff
@@ -797,3 +797,39 @@ theorem riemannZeta_neg_nat_eq_bernoulli (k : ℕ) :
     rw [inv_mul_cancel (ofReal_ne_zero.mpr <| pow_ne_zero _ pi_pos.ne'),
       inv_mul_cancel (ofReal_ne_zero.mpr <| pow_ne_zero _ two_ne_zero), one_mul, one_mul]
 #align riemann_zeta_neg_nat_eq_bernoulli riemannZeta_neg_nat_eq_bernoulli
+
+/-- The residue of `Λ(s)` at `s = 1` is equal to 1. -/
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/ZetaFunction.lean:801]: You may need to `Open Topology` for the neighborhood notation.

File: Mathlib/NumberTheory/ZetaFunction.lean [hunk 2]
```diff
@@ -797,3 +797,39 @@ theorem riemannZeta_neg_nat_eq_bernoulli (k : ℕ) :
...
+lemma riemannCompletedZeta_residue_one :
+    Tendsto (fun s ↦ (s - 1) * riemannCompletedZeta s) (𝓝[≠] 1) (𝓝 1) := by
+  unfold riemannCompletedZeta
+  simp_rw [mul_add, mul_sub, (by simp : 𝓝 (1 : ℂ) = 𝓝 (0 - 0 + 1))]
+  refine ((Tendsto.sub ?_ ?_).mono_left nhdsWithin_le_nhds).add ?_
+  · rw [(by simp : 𝓝 (0 : ℂ) = 𝓝 ((1 - 1) * riemannCompletedZeta₀ 1))]
+    apply ((continuous_sub_right _).mul differentiable_completed_zeta₀.continuous).tendsto
+  · rw [(by simp : 𝓝 (0 : ℂ) = 𝓝 ((1 - 1)  * (1 / 1)))]
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/ZetaFunction.lean:809]:
```suggestion
  · rw [(by simp : 𝓝 (0 : ℂ) = 𝓝 ((1 - 1) * (1 / 1)))]
```

File: Mathlib/NumberTheory/ZetaFunction.lean [hunk 3]
```diff
@@ -797,3 +797,39 @@ theorem riemannZeta_neg_nat_eq_bernoulli (k : ℕ) :
...
+      continuousAt_const.div continuousAt_id one_ne_zero)
+  · refine (tendsto_const_nhds.mono_left nhdsWithin_le_nhds).congr' ?_
+    refine eventually_nhdsWithin_of_forall (fun s hs ↦ ?_)
+    simpa using (div_self <| sub_ne_zero_of_ne <| Set.mem_compl_singleton_iff.mpr hs).symm
+
+/-- The residue of `ζ(s)` at `s = 1` is equal to 1. -/
+lemma riemannZeta_residue_one : Tendsto (fun s ↦ (s - 1) * riemannZeta s) (𝓝[≠] 1) (𝓝 1) := by
+  suffices : Tendsto (fun s => (s - 1) *
+      (π ^ (s / 2) * riemannCompletedZeta s / Gamma (s / 2))) (𝓝[≠] 1) (𝓝 1)
+  · refine this.congr' <| (eventually_ne_nhdsWithin (one_ne_zero' ℂ)).mp (eventually_of_forall ?_)
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/ZetaFunction.lean:820]: Slightly shorter:
```suggestion
  · refine this.congr' <| (eventually_ne_nhdsWithin one_ne_zero).mp (eventually_of_forall ?_)
```

File: Mathlib/NumberTheory/ZetaFunction.lean (hunk 1 above)
Comment (loefflerd) [reply Mathlib/NumberTheory/ZetaFunction.lean:801]: I don't think so – there's `open scoped Topology` at the head of the file. It compiles ok on my machine, so there shouldn't be any imports missing.

File: Mathlib/NumberTheory/ZetaFunction.lean (hunk 3 above)
Comment (loefflerd) [reply Mathlib/NumberTheory/ZetaFunction.lean:820]: 🏌️

File: Mathlib/NumberTheory/ZetaFunction.lean [hunk 4]
```diff
@@ -797,3 +797,39 @@ theorem riemannZeta_neg_nat_eq_bernoulli (k : ℕ) :
...
+  have h1 : Tendsto (fun s : ℂ ↦ 1 / Gamma (s / 2)) (𝓝[≠] 1) (𝓝 (1 / π ^ (1 / 2 : ℂ)))
+  · rw [← Complex.Gamma_one_half_eq]
+    refine (Continuous.tendsto ?_ _).mono_left nhdsWithin_le_nhds
+    simpa using differentiable_one_div_Gamma.continuous.comp (continuous_id.div_const _)
+  convert (riemannCompletedZeta_residue_one.mul h0).mul h1 using 2 with s
+  · ring
+  · rw [one_mul, mul_one_div, div_self]
+    rw [(by simp : (1 / 2 : ℂ) = ↑(1 / 2 : ℝ)), ← ofReal_cpow pi_pos.le, Ne.def, ofReal_eq_zero,
+      rpow_eq_zero_iff_of_nonneg pi_pos.le, not_and_or]
+    exact Or.inl pi_ne_zero
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/ZetaFunction.lean:835]:
```suggestion
  · have : (π : ℂ) ^ (1 / 2 : ℂ) ≠ 0 :=
      fun h ↦ ofReal_ne_zero.mpr pi_ne_zero ((cpow_eq_zero_iff _ _).mp h).1
    field_simp
```

File: Mathlib/NumberTheory/ZetaFunction.lean (hunk 1 above)
Comment (MichaelStollBayreuth) [reply Mathlib/NumberTheory/ZetaFunction.lean:801]: OK. I copied the new stuff into a test file opening what `#where` in the original file said, but apparently `#where` does not show `Open scoped ...` stuff.

File: Mathlib/NumberTheory/ZetaFunction.lean [hunk 5]
```diff
@@ -797,3 +797,37 @@ theorem riemannZeta_neg_nat_eq_bernoulli (k : ℕ) :
...
+  have h0 : Tendsto (fun s ↦ π ^ (s / 2) : ℂ → ℂ) (𝓝[≠] 1) (𝓝 (π ^ (1 / 2 : ℂ)))
+  · refine ((continuousAt_id.div_const _).const_cpow ?_).tendsto.mono_left nhdsWithin_le_nhds
+    exact Or.inl <| ofReal_ne_zero.mpr pi_ne_zero
+  have h1 : Tendsto (fun s : ℂ ↦ 1 / Gamma (s / 2)) (𝓝[≠] 1) (𝓝 (1 / π ^ (1 / 2 : ℂ)))
+  · rw [← Complex.Gamma_one_half_eq]
+    refine (Continuous.tendsto ?_ _).mono_left nhdsWithin_le_nhds
+    simpa using differentiable_one_div_Gamma.continuous.comp (continuous_id.div_const _)
+  convert (riemannCompletedZeta_residue_one.mul h0).mul h1 using 2 with s
+  · ring
+  · haveI := fun h ↦ ofReal_ne_zero.mpr pi_ne_zero ((cpow_eq_zero_iff _ (1 / 2)).mp h).1
```
Comment (alexjbest) [inline Mathlib/NumberTheory/ZetaFunction.lean:832]:
```suggestion
  · have := fun h ↦ ofReal_ne_zero.mpr pi_ne_zero ((cpow_eq_zero_iff _ (1 / 2)).mp h).1
```
haveI means have + inline, it should basically never be necessary in proofs

Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/ZetaFunction.lean:832]: I don't think you need `haveI` here. (My understanding is that `I` stands for "inline", but there is no occurrence of `this` that would make sense being inlined.)
```suggestion
  · have := fun h ↦ ofReal_ne_zero.mpr pi_ne_zero ((cpow_eq_zero_iff _ (1 / 2)).mp h).1
```

Comment (loefflerd) [reply Mathlib/NumberTheory/ZetaFunction.lean:832]: OK, I had some vague notion that it was slightly faster to compile than regular `have` (but even if so, the speed gain will clearly be barely noticeable in this proof). Happy to take it out again.

**Review by MichaelStollBayreuth — APPROVED** [review]
LGTM!

File: Mathlib/NumberTheory/ZetaFunction.lean (hunk 5 above)
Comment (alexjbest) [reply Mathlib/NumberTheory/ZetaFunction.lean]: If you find a compelling example of this please report it on the zulip, I imagine people would be interested in trying to understand if something is at fault, as far as I understand there shouldn't be a big difference (but of course my understanding could be the thing at fault!)

**Comment by alexjbest** [comment]
maintainer merge

**Comment by loefflerd** [comment]
Thanks for the reviews Michael and Alex!

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**